### ISSUE TITLE
Feature: Collaborative editing (at least conflict/locking)

### ISSUE THREAD

**Issue by bios-ben** [BookStackApp/BookStack#395]
Desired Feature:
Currently if you have multiple people editing the same document, the most recent person to save just overwrites the other. At the very least this should use lock files and note when there's conflicts or try to auto merge. But the better feature would be to collaborative edit using sockets like other editing tools such as etherpad or hackmd.

**Comment by Shackelford-Arden** [comment]
While I +1 this, I'd also say that it would be a nice addition to have the option between Check In/Out method and collaborative writing. Some may not want the collaborative ability. Personally I would, but I know my last job would simply want a check in/out functionality.

**Comment by ssddanbrown** [comment]
Yeah, I can see the benefit of something like this. Collaboration will be hard to do technically, Especially on the current laravel codebase without the app structure getting extra complicated. My concerns with locking is that it kind of relies on the user to lock/unlock which we can the work around with notifications and such but it does add a lot to the system.

At the moment the system will notify upon edit if someone else has started editing but there's nothing forceful in place.

**Comment by therealscottcarlow** [comment]
Locking or check-in/check-out is going to be necessary for my org at scale. I would love to see this implemented. Collab is OK, but realistically if I wanted to collab on a doc I wouldn't use Bookstack as the platform to do that in. That's my 2 cents.

**Comment by bios-ben** [comment]
What I'd like to see in an app, if I were to make something ideal from scratch, is to be able to edit a file and choose collaborative, real-time read-only, or private lock. In my org the main thing I want a lot of the time is to have multiple people live viewing the edits as they go in (in final formatted form), while one person is doing the editing using whatever editor (markup, wysiwyg, etc). This allows people to essentially audit the documenting party, while only one person is responsible for actual dictation. I haven't seen software like this, and if I had the cycles/funding I'd build it. 

----- Original Message -----

**Comment by dsaundersWizmo** [comment]
+1 for Check In/Check Out.

Would also need admin level force Check In and probably a way to do that in bulk.  Check Out expiration would be nice.

**Comment by vdavid** [comment]
Re: @ssddanbrown

> At the moment the system will notify upon edit if someone else has started editing but there's nothing forceful in place.

Actually, that'd be cool but it doesn't seem to do that.
I tested it on the demo site: I logged in with `admin`, created a `guest2` user, logged in as `guest2` in a porn window, started editing a page as `admin`, but when I went there as `guest2` there was no indication whatsoever that the page was being edited. I don't know if I should report this as a bug, because I'm not sure if it's broken or just nonexistent.

**Comment by NickeZ** [comment]
I have been using HackMD for a while and got really used to collaborative editing. The use cases I see are for example writing meeting minutes and doing project plan drafts which are easier done as a team effort if everyone can at least live-view the updates (preferably also live edit).

**Comment by foinus** [comment]
we also need this and it shut also prevent moving pages when somebody is editing them

**Comment by alexanderrieder** [comment]
we would also appreciate if others at least could not edit the same page while another user is editing it :)

**Comment by Valiantiam** [comment]
I haven't seen anyone post here for this request in a while. This is a must for any medium sized teams attempting to work on documentation at the same time to prevent conflicts and lost data and time.

**Comment by bios-ben** [comment]
Fun fact, CKEditor 5 provides this feature out of the box now and is quite polished.

**Comment by ssddanbrown** [comment]
I don't believe the collaboration components of CKEditor5 are open-source and, Even then, it would still require some considerable wiring to the back-end.

I've been keeping an eye on [prosemirror](https://prosemirror.net/) as a potential way to combine the WYSIWYG & Markdown editors and at least provide some kind of route to collaboration. Lots of potential work and issues to journey that path though.

**Comment by Valiantiam** [comment]
That looks somewhat promising. Again, Ideally, even if we just had a way to "lock" editing to one user at a time, so information is not overwritten by two people trying to write to the same document/page etc.

**Comment by Rdnaskel** [comment]
Is there any kind of a roadmap or at least a plan to implement collab editing at this moment? @ssddanbrown have mentioned prosemirror which looks nice but is there something devs are doing to make this happen? 🙂

**Comment by kartha10** [comment]
+1

**Comment by farfalk** [comment]
Hi everyone,
we're using bookstack in our infrastructure. This feature would be extremely useful, especially with an hackmd-like ux. I'd like to know, if it's possible of course, how could I as a programmer / we as a company help in implementing it or accelerate its development.

**Comment by jstiops** [comment]
hi, how are things going with this?

**Comment by Hazzard17h** [comment]
Maybe you could be interested in this pull request: https://github.com/BookStackApp/BookStack/pull/1862
It is based on version 0.29.3, I had not rebased it afterwards.

**Comment by OkoGoran** [comment]
Hello there! I also wonder if there is any plan to integrate this feature somehow. At least a warning is a killer-criteria. Otherwise one of two people who are working on the same page will loose all it's content, without even noticing it.

**Comment by Valiantiam** [comment]
Yeah considering the response [here](https://github.com/BookStackApp/BookStack/pull/1862#issuecomment-598456070) I understand not wanting to add collaboration.

But as @OkoGoran said, we do need then some sort of warning that someone else is editing, or even a lock period, where a page is locked when someone else is editing. This would also introduce a couple other issues that would need to be addressed. Such as if you want to access a page someone kept open on accident... you'd need a way to time that other person out, or allow the next person to force the other persons work to save, and close, so you can keep edition. There are a few different ways to tackle that issue I'd imagine.

**Comment by ssddanbrown** [comment]
@Valiantiam Just to confirm, my comments there were in regards to sharing the existing draft system. Collaborative editing is different and something I would like one day, it's just a lot tricker to implement.

> At least a warning is a killer-criteria

@OkoGoran You should get a warning if you start editing a page that another user has started editing (Or saved a draft for) in the last hour. You should also get a warning if you start editing one of your drafts where the page has since been saved.

> Otherwise one of two people who are working on the same page will loose all it's content, without even noticing it.

@OkoGoran It can be hard to notice overwrites, But the content should not be lost. If two people save the same page with different content, you should still have the content for both saved as revisions so you can always see/revert as required.

---

Just to confirm the status of things, A review of alternate page editors is upcoming on the roadmap. Collaborative editing compatibility will be considered during that phase. There are other technical/support hurdles along the way, but the editor is one of the biggest.

**Comment by tmanok** [comment]
1+ to making this happen. Confluence Server just fell off a cliff, I really need an alternative and Book Stack was under consideration a while back. I'm hoping that I can migrate over a hundred people somewhere and they need collaborative editing. I would also consider hiring a developer next year to support any open source projects that the company uses, including Bookstack if things turn out. 
Thanks!

**Comment by nikkilocke** [comment]
+1 to this. 

Just to throw some wild ideas into the mix 
- could you leverage git somehow for merging changes when two people edit at the same time?
- Nextcloud has collaborative editing of markdown files, which works pretty well

If user A edits something, then user B edits the same thing at the same time. 
- When they save, which version becomes the live page - is it the most recent save?
- Do both versions make it into the history?

**Comment by christoferw** [comment]
i think we should use some kind of websockets, e.g.
https://beyondco.de/docs/laravel-websockets/getting-started/introduction to
enable real time coloboration


Viele Grüße aus Coesfeld

Christofer

________________________
*Dipl. Ing. Christofer Wesseling*
Geschäftsführer
02541 / 880 - 117
cw@weslink.de


*WESLINK GmbH*
Goxel 33b
48653 Coesfeld
www.weslink.de

eBusiness Softwareentwicklung
Frontend | Backend | Consulting


Sitz der Gesellschaft: Coesfeld | Amtsgericht Coesfeld HRB 17888 |
Geschäftsführer: Dipl. Ing. Christofer Wesseling
DE: Diese E-Mail enthält vertrauliche und/oder rechtlich geschützte
Informationen. Wenn Sie nicht der richtige Adressat sind oder diese E-mail
irrtümlich erhalten haben, informieren Sie bitte sofort den Absender und
vernichten Sie diese Mail. Das unerlaubte Kopieren sowie die unbefugte
Weitergabe dieser Mail ist nicht gestattet. EN: This e-mail may contain
confidential and/or privileged information. If you are not the intended
recipient (or have received this e-mail in error) please notify the sender
immediately and destroy this e-mail. Any unauthorized copying, disclosure
or distribution of the material in this e-mail is strictly forbidden.


Am Fr., 5. Feb. 2021 um 15:45 Uhr schrieb Nikki Locke <
notifications@github.com>:

**Comment by tmanok** [comment]
> +1 to this.
> 
> Just to throw some wild ideas into the mix
> 
> * could you leverage git somehow for merging changes when two people edit at the same time?
> * Nextcloud has collaborative editing of markdown files, which works pretty well
> 
> If user A edits something, then user B edits the same thing at the same time.
> 
> * When they save, which version becomes the live page - is it the most recent save?
> * Do both versions make it into the history?

Hey Nikkilocke,

I think that being able to edit documents together collaboratively is the goal. Being able to see text the moment another person adds it to the document in the document editor is very important feature. If you've never encountered this before, consider the collaborative editing found in Google Docs or Confluence. There is an indication of who is currently viewing the document and all parties can see what changes are being made live. This all depends on the permissions for the document however, inheriting space permissions would mean that new documents are visible to any permitted users of the space, but that could be easily changed by the document's owner. 

I like Christoferw's suggestion of using websockets, though I'm not technically proficient enough to contribute code for it. 

Thanks,

Triston

**Comment by Redsandro** [comment]
@ssddanbrown commented:

> @OkoGoran You should get a warning if you start editing a page that another user has started editing (Or saved a draft for) in the last hour.

I don't see this when editing with two different users on two different incognito windows of Google Chrome using __BookStack__ `21.04.3` for testing purposes.

My BookStack is set to Markdown mode, not WYSIWYG mode. Is that relevant?

---

Google brought me here. I have this problem where someone overwrote someone elses contribution. Only happened once, but I've only been using BookStack for a week.

**Comment by bamhm182** [comment]
I'm only seeing it sometimes. I created a page and made edits, then I saved it and started editing it again. I had a team member try to edit it and he did not see the popup. He saved his work and I saved mine. My work clobbered his work. I went to edit again, and it told me that he had edited it.

I would really prefer to see an option that would hard lock a page if there's a draft.

**Comment by MatthieuParis** [comment]
You see a warning when opening the editor only if the other user editing the same page has already saved a draft.

Here is a pull request adding a warning each time a draft is saved: #2877.

**Comment by Atn-D** [comment]
Hello, I was looking at the CKeditor5 website and realized that it is possible to implement real-time collaboration!

If it can give some tracks for this issue.

https://ckeditor.com/collaboration/real-time-collaborative-editing/

**Comment by ssddanbrown** [comment]
@Atn-D I'm not sure the real-time collaboration components and protocol are fully open source. Even if it was, The CKeditor5 licence is not compatible for our usage. That's ignoring the general effort and compatibility considerations with switching from the existing editor.

**Comment by HackingWolpertinger** [comment]
without wanting to get on anyone's nerves, is there any news yet on the plans for collaborative editing in bookstack? Are there plans to include, for example, an Etherpad-like solution? I think bookstack is cool and that would make it even more practical in my eyes.

**Comment by benjamindonze** [comment]
Hello +1 on this. For me it's almost a showstopper that 2 users can actually change a page and overwrite each other with no error message. It makes collaboration almost not possible. We are currently selecting a wiki tools and for us this is possibly going to force us to another solution.

At very least optimistic locking should be implemented so that if 2 users edit at same time the first one who save success but the second one fail with error message informing that another version was created in the mean time.

**Comment by mi-wulf** [comment]
@ssddanbrown Are there still plans to create at least a lock feature while editing? Otherwise people are constantly overwriting each others content. While testing I mentioned you get a message, but I dont think that is enough(and I think you get the first message after a minute when the auto save draft kicks in), for example in most CMS/DXP systems you get a message box if someone else already working on a draft and you can decide if you want to edit anyway or discard.

If there are no plans to implement a lock system or a collab system is there a recommendation how to implement a lock system?

**Comment by ssddanbrown** [comment]
@mi-wulf I've never had any solid plans tbh. As I alluded to in my original, now half-decade old :fearful: comment, I'm not really a fan of locking and that user flow of unlocking/locking, it feels somewhat corporate, putting process over UX which I'd want to avoid, upon the maintenance of those systems. 

From my point of view, I wouldn't want to add specific process/features for this right now since, based upon user feedback and desires over the years, I see us longer term moving towards a more significant/formal draft/publish flow which will change up and drive requirements in this area.

For the shorter-term, If this is genuinely an active problem right now though we could instead look to make further changes to add a little more friction in these scenarios. I'm thinking along the lines of:
- Require active continue confirmation upon dual editing rather than just a passive message.
- Create an early draft upon edit so that another "active editing session" is considered from point-of-edit rather than ~30s after first changes made.

Happy to hear feedback in regards to whether those would help, or other potential easy-wins.

---

Thought I'd address my [2021 comment regarding collaboration](https://github.com/BookStackApp/BookStack/issues/395#issuecomment-762392131 ) while I'm here. The summary is that the mentioned editor review was a bad time and didn't go in the direction I wanted, which has since dampened my spirits in regards to live collaboration.

**Comment by benjamindonze** [comment]
As said above a very simple and minimal implementation would be optimistic locking.

When you save the modification on a page you also send the timestamp (or any version ID) of the version you based your modification on. If another user has modified the page in the meantime (you can check that based on version ID) then an error message is displayed saying the page was modified in between. From there you can maybe offer to user either to cancel or to override the changes but at least user is informed.

**Comment by mi-wulf** [comment]
@ssddanbrown 
While I prefer stricter locking system where the you get a locked message and you can only create a new draft if the active drafts are discarted or published by the user currently editing you mentioned features would help a lot to advance in this topic.

> Require active continue confirmation upon dual editing rather than just a passive message.

_Here a message box like @benjamindonze mentioned would be nice. In the message box that pops up should be the information who is currently editing(has the open draft) and the age of the existing draft, something like "There is currently a draft open from user <username>, the draft was created <date>." Then I have the possability to get in contact with that person.

> Create an early draft upon edit so that another "active editing session" is considered from point-of-edit rather than ~30s after first changes made.

_So here the draft should be created by clicking on edit and maybe use the way @benjamindonze explained with the date._

**Comment by mi-wulf** [comment]
@ssddanbrown Any idear when something like that could be implemented?

**Comment by fitterfizzle** [comment]
@ssddanbrown has there been any progress on this?

**Comment by LHBL2003** [comment]
We also have the problem that sometimes 2 people work on it. We would also be happy about this function.

**Comment by LHBL2003** [comment]
> Yeah, I can see the benefit of something like this. Collaboration will be hard to do technically, Especially on the current laravel codebase without the app structure getting extra complicated. My concerns with locking is that it kind of relies on the user to lock/unlock which we can the work around with notifications and such but it does add a lot to the system.
> 
> At the moment the system will notify upon edit if someone else has started editing but there's nothing forceful in place.

Is there a notification that someone else is also working on the same page? We don't see anything anywhere. But either I'm misunderstanding the post or there used to be this function?

It would be enough for us to know that someone else is also currently in edit mode.

**Comment by dmoebius** [comment]
@ssddanbrown Since #2877 is in place for a couple of years now, wouldn't it be the easiest just to save a draft immediately as soon as the user starts to edit a page, ie. goes into edit mode?!? This way he participates in the messaging implemented in #2877 and gets warning messages right from the start.

**Comment by kobaz** [comment]
https://etherpad.org/
Etherpad is a highly customizable open source online editor providing collaborative editing in really real-time.

**Comment by berils** [comment]
Its pretty common, that two colleagues using same page, and making changes. Can we wait for some sort of solution? (online editin, locked file, etc.)
BR,
Berils.

**Comment by benjamindonze** [comment]
Anything going on regarding this?

**Comment by bendem** [comment]
> It can be hard to notice overwrites, But the content should not be lost. If two people save the same page with different content, you should still have the content for both saved as revisions so you can always see/revert as required.

Technically, your database will handle this for with transactions. When you load the editor, load the current revision number of the page in a hidden field, when saving, check if the latest revision still matches. Since you check the latest revision, save the changes, and insert the new revision in the same transaction, that operation is atomic and cannot fail, the last person to save will fail the check and can be sent back to the editor with an "error" (or a diff view, but that's probably a lot harder).

That method is what most application implement regarding content overwrite. It's a battle tested technique and would solve the current silent overwrite.

If that's not feasible, I'd suggest splitting this issue in two, collaboration is a totally different subject than locking (even if they are somewhat exclusive).

**Comment by benjamindonze** [comment]
> > It can be hard to notice overwrites, But the content should not be lost. If two people save the same page with different content, you should still have the content for both saved as revisions so you can always see/revert as required.
> 
> Technically, your database will handle this for with transactions. When you load the editor, load the current revision number of the page in a hidden field, when saving, check if the latest revision still matches. Since you check the latest revision, save the changes, and insert the new revision in the same transaction, that operation is atomic and cannot fail, the last person to save will fail the check and can be sent back to the editor with an "error" (or a diff view, but that's probably a lot harder).
> 
> That method is what most application implement regarding content overwrite. It's a battle tested technique and would solve the current silent overwrite.
> 
> If that's not feasible, I'd suggest splitting this issue in two, collaboration is a totally different subject than locking (even if they are somewhat exclusive).

I agree that it should be splitted in two feature. Most important would first to solve the silent concurrent update issue as proposed with optimistic locking. 
Then collaboration could be another feature.

**Comment by BloodyIron** [comment]
@ssddanbrown 

I am not entirely 100% up to speed on where this topic is within bookstack, but I want to lend some thoughts that hopefully can turn into something productive.

From what I understand, one of the more tangible concerns on this greater topic is the loss of data between people working on the same document at the same time. That's a very tangible thing and I would dare say we can all agree is legitimate.

There may be other things to this too, but I don't necessarily want to be exhaustive in coverage of the matter at this time.

Something that is perhaps worth for consideration to address this is collaboration of the same document in a shared session so to say. In that, whenever a document is edited, all users that want to edit said document edit the same document session at the same time.

I know of two examples that I'm pretty sure are really great examples of how this _could_ work. Namely within nextCloud:

1. Draw.io integration
2. Nextcloud Office (in my case the "CODE" Docker image/container)

For Draw.io, whenever I open a diagram file, I'll see the mouse cursor of any other user (even myself) that also has that same file open. It's actually a bit spooookkkyyy to see my own cursor when I open the same doc up in another window, forgetting I have it literally visible on another monitor, hah! But either way, this is a good example of live collaboration.

For the Nextcloud Office example, you'll see when other users are working on the same document. Where their cursor (as in typing cursor, not mouse cursor) is, will be highlighted with a different colour per user/session, and whatever each person does to the document does not actually interrupt each other.

To me these are both fabulous examples of how collaboration of this nature can be successfully done. I can't speak to the technological mechanisms that make this possible, but I do highly recommend these examples be considered for possible methods that could be used (at least in concept) within Bookstack.

**Comment by kobaz** [comment]
> @ssddanbrown
> 
> I am not entirely 100% up to speed on where this topic is within bookstack, but I want to lend some thoughts that hopefully can turn into something productive.
> 
> From what I understand, one of the more tangible concerns on this greater topic is the loss of data between people working on the same document at the same time. That's a very tangible thing and I would dare say we can all agree is legitimate. ................



Great writeup here.

In terms of the tech, looks like Etherpad would be capable of doing just that.  The good news is that it looks also like the MIT License and Apache License are compatible.

I completely agree on the default behavior to just be jumping straight into a collaborative editing session.  If someone wants to privately edit their own version however, the existing drafts system would accommodate.

**Comment by BloodyIron** [comment]
> > @ssddanbrown
> > I am not entirely 100% up to speed on where this topic is within bookstack, but I want to lend some thoughts that hopefully can turn into something productive.
> > From what I understand, one of the more tangible concerns on this greater topic is the loss of data between people working on the same document at the same time. That's a very tangible thing and I would dare say we can all agree is legitimate. ................
> 
> Great writeup here.
> 
> In terms of the tech, looks like Etherpad would be capable of doing just that. The good news is that it looks also like the MIT License and Apache License are compatible.
> 
> I completely agree on the default behavior to just be jumping straight into a collaborative editing session. If someone wants to privately edit their own version however, the existing drafts system would accommodate.

To expand on the thought of "private" edit sessions. If we were to reach a position where that functionality is possible (private edit sessions), it might be worthwhile as a setting to allow/disallow shared editing and/or private editing. That way different environments can tune what is on/off to meet their environmental needs. Some environments might want ONLY shared editing on, and no private editing. Some environments might want ONLY private editing. Some might want both! (or none? no editing for you! guess I found a gotcha haha)

**Comment by pnocera** [comment]
Coming here because I was searching for a feature like "Merge Track Changes" in BookStack. 
Real time collaborative editing is not exactly the point. The feature I need is rather having several people work on their own copy, then have one responsible person to review / accept / reject changes.
Sounds like git doesn't it ?
Before going into coding, I'm wondering if there are already projects or initiatives to integrate BookStack with Git ( eg Gitea ) ?

**Comment by HelloGithubUniverse** [comment]
We run into the Same problem … When two Teammates work on the same document without knowing that somebody else is working on it. Wouldnt it be a simple solution instead of live collaboration like etherpad just to set a database entry like  „locked“ for the page?

**Comment by bendem** [comment]
I proposed a change to at least prevent data loss, it's far from perfect and probably not up to the project standards, but it shows at least that it's feasible.

**Comment by sebhahn** [comment]
Any progress on this issue?

**Comment by jurrevanderschaaf** [comment]
We need to migrate away from Confluence and BookStack is our nr 1 option! However, this issue is currently blocking the migration.

A locking system where only one user can edit a page at a time would do it for us! Just want to avoid users losing their changes. It may not be a solution for everyone, so maybe this could be an optional setting?

**Comment by deanord** [comment]
We were just about to drop Confluence and switch when someone pointed this out.  Collaborative editing would be the preferred option, but locking would be acceptable.  Writing over someone else's edits is a no-go.  This makes Bookstack a single-user app, IMO.  I love this app, but this is a big deal!

**Comment by Maarten-TL** [comment]
Would also like to upvote some form of collaborative editing (but a bit simpler than full live-sharing) to accommodate a simple approval flow that we need. In our case, even if the collaborative editing is 'turn by turn' it will be fine. The bottleneck for us is that drafts that are saved are currently individual with (AFAIK) no way to let one user save a draft that another user can then see or continue to work on.

The use-case for us is to implement the 'four eyes' principle before accepting changes. This means that changes are made as 'draft' and don't overwrite the main page, but these draft-changes can be seen by someone other than the author of the draft, who can approve-and-publish. To keep things simple, we don't require the tool to enforce the workflow, but just to accommodate it as a possibility.

Since we make limited changes and usually in a team of just a few people (the author and the reviewer/approver) any kind of 'live collaboration' is not so important, but it will be useful if the people can see and continue work on the same draft (in turns) before deciding to publish the result.

This also allows many small page edits (by several people) to be collected over time, for one overarching approval from time to time.

The request shares some similarities with the feature request for workflow #473 but I felt it more closely fits the discussion on collaborative editing as we don't require enforcing the workflow necessarily.

As indicated at the start, I feel that the change needed for this can be relatively small: Just store drafts for a page 'globally' instead of 'per user'. (I understand supporting features long term can be an issue and there may be other reasons not to do it - but I hope that if indeed this is a relatively small change / configuration option that can be added, it can help many people in similar situations).

(edit: perhaps to add: We have been using bookstack for about 2 years and I'm looking to see if we can expand its use internally for more purposes)

**Comment by intellasoft** [comment]
I think a simple websocket stream that updates viewers of a locking status would accomplish this.

Clients<->Server -- Ongoing WebSocket for events

Client1 -> Requests Edit
Server->Broadcast Lock
Client2 <- Receives Lock
Server->Grants Edit to Client1
Client1-> Saves Draft
Client2 <- Receives updated Draft
Client2 <- Still locked from Edit (But can of course make a page copy and edit a second page to avoid blocking)
Client1 -> As long as Client1 is in edit mode, Other clients are locked from editing drafts or publishing


Endpoint cases:

1)
Client1 WebSocket times out/disconnects
Server->Broadcast unlock
Client2 can now edit the last saved draft made by Client1
Client2 can now edit the page

2)
Client1 Publish
Server->Broadcast unlock
Client2 <- Receives updated publish
Client2 can now edit the page


You could get fancy and add lock overrides (steal session/lock, and then Lock out Client1 from his session), which would force Client1 to reload the page or at least pop up a dialog saying the page session has been stolen and their page is now out of date/etc

**Comment by intellasoft** [comment]
@Maarten-TL 
In reply to
> The use-case for us is to implement the 'four eyes' principle before accepting changes. This means that changes are made as 'draft' and don't overwrite the main page, but these draft-changes can be seen by someone other than the author of the draft, who can approve-and-publish. To keep things simple, we don't require the tool to enforce the workflow, but just to accommodate it as a possibility.

What about a workaround of a 'Drafts' shelf that has no permission for view by the public, where a fully copy is stored before a publish occurs whereby you copy/overwrite the existing with the approved new version from the Drafts shelf

**Comment by intellasoft** [comment]
My brainstorming has also thought of:
- If you're building in a websocket for collaboration, one could implement 'poor mans collaborative editing' by
1) Streaming all edits to other watching clients via events
2) Store cursor position, Force their page to reload
3) 'Other Clients' now see the new edits
4) Repeat the cycle

Of course this also hinges on the ability for users to see each other's drafts

**Comment by benalbrecht** [comment]
i've been evaluating bookstack for our organization and i've also stumbled upon the lack of collaborative editing.

fortunately the biggest prerequisite (support in the editor) is already done via the migration to the new wysiwyg editor based on lexical https://github.com/BookStackApp/BookStack/issues/5631 which in turn supports [yjs](https://yjs.dev/) for state syncing.

i built a proof of concept that uses [hocuspocus](https://github.com/ueberdosis/hocuspocus) as a websocket sync server.
the syncing seems to work reliably across multiple browser windows.
on the bookstack (php) side, it's integrated via a db table that keeps track of editing sessions and an endpoint that handles auth and join/leave events, about ~300loc backend code

obviously, a separate nodejs based sync service would complicate the deployment story, but as an optional feature that can be enabled on demand, it might be an acceptable trade-off.

@ssddanbrown i could put together a draft PR over the next few days, if this is a direction that you'd consider. i didn't find a php library with a comparable feature set.

**Comment by ssddanbrown** [comment]
> i could put together a draft PR over the next few days, if this is a direction that you'd consider. 

Thanks @benalbrecht for offering that, but that's not a route I'd want to take just yet.

> fortunately the biggest prerequisite (support in the editor) is already done via the migration to the new wysiwyg editor based on lexical

Yeah, the move to the new editor was with this in mind, so we have a nice base data model to work from, and yeah, I've kept the YJS bindings in our fork, ready for potential use.

I did play with this last year, but got a little burnt out while attempting to rewrite yjs as a PHP library.
I don't want to give up on a PHP based solution right yet, and I hope to try again this year (Maybe also explore a potential PHP adapter to the rust port).
Maybe if that falls flat again I'll give up and resort to having a separate component as per your approach.

The data merging is one of two remaining big parts in my view, with the other being a bundle of all the edge-cases and process handling to support this while remaining compatible with existing flows/editors/use-cases.

**Comment by jurrevanderschaaf** [comment]
Honestly, we'd love to deploy a separate sync service, like hocuspocus or something else, to have such a feature possible. Also if this is temporary. It can be replaced later by a PHP based solution. I understand it is desirable to have a PHP based solution. But for us, a more complicated deployment setup is worth it to have such a feature.

**Comment by dwt** [comment]
Then way not start with either 
- on starting the edit: show that page is currently edited and allow the user to override. No complication in setup required.

- on saving: show that the page was edited and show a diff (like Wikipedia used to do) and the edit field below for the user to reconcile

**Comment by ssddanbrown** [comment]
BookStack is now [managed on Codeberg](https://codeberg.org/bookstack/bookstack).

As part of this change, we're closing all existing GitHub issues and pull requests.
You'll instead find this issue on Codeberg here: 
https://codeberg.org/bookstack/bookstack/issues/395
  
Any further discussion for this will now be done on Codeberg.
Therefore this GitHub thread will be locked.
